File: src/main/java/com/thomaskint/minidao/model/MDCallStack.java
package com.thomaskint.minidao.model;

import java.util.ArrayList;
import java.util.List;

public class MDCallStack<T> {

	private List<T> stack;

	public MDCallStack() {
		this.stack = new ArrayList<>();
	}

	public boolean isEmpty() {
		return this.stack.isEmpty();
	}

	public T push(T item) {
		this.stack.add(item);
		return item;
	}

	public T pop() {
		T item = this.stack.get(this.stack.size() - 1);
		this.stack.remove(item);
		return item;
	}

	public int search(T item) {
		int index = -1;
		int i = 0;
		while (i < stack.size() && index == -1) {
			if (stack.get(i).equals(item)) {
				index = i;
			}
			i++;
		}
		return index;
	}

	public boolean isPresent(T item) {
		return search(item) != -1;
	}
}

File: src/main/kotlin/com/thomaskint/minidao/model/MDCallStack.kt
package com.thomaskint.minidao.model

import java.util.ArrayList

class MDCallStack<T> {

    private val stack: MutableList<T>

    val isEmpty: Boolean
        get() = this.stack.isEmpty()

    init {
        this.stack = ArrayList()
    }

    fun push(item: T): T {
        this.stack.add(item)
        return item
    }

    fun pop(): T {
        val item = this.stack[this.stack.size - 1]
        this.stack.remove(item)
        return item
    }

    fun search(item: T): Int {
        var index = -1
        var i = 0
        while (i < stack.size && index == -1) {
            if (stack[i] == item) {
                index = i
            }
            i++
        }
        return index
    }

    fun isPresent(item: T): Boolean {
        return search(item) != -1
    }
}
